modules/dnn/src/op_cuda.hpp

```diff
@@ -0,0 +1,358 @@
+                CV_Assert(mat.type() == CV_32F);
+                CV_Assert(mat.total() >= tensor.size());
+
+                Mat source = mat.isContinuous() ? mat : mat.clone();
+                CV_Assert(source.isContinuous());
+
+                memcpy<float>(tensor.get(), reinterpret_cast<float*>(source.data), tensor.size(), stream);
+            }
+
+            /** @brief copies data from a TensorType to a cv::Mat
+             *
+             * Pre-conditions:
+             * - \p mat must be larger or equal to the tensor in size
+             *
+             * @note performance is best for continuous and page-locked cv::Mat
+             */
+            template <class T>
+            void copyTensorToMat(Mat& mat, TensorView<T> tensor, const Stream& stream);
+
+            template <> inline
+            void copyTensorToMat(Mat& mat, TensorView<half> tensor, const Stream& stream)
+            {
+                CV_Assert(mat.type() == CV_32F);
+                CV_Assert(mat.total() >= tensor.size());
+
+                Mat source(shape(mat), CV_16F);
+                CV_Assert(source.isContinuous());
+
+                memcpy<half>(reinterpret_cast<half*>(source.data), tensor.get(), tensor.size(), stream);
+
+                source.convertTo(mat, CV_32F);
+            }
+
+            template <> inline
+            void copyTensorToMat(Mat& mat, TensorView<float> tensor, const Stream& stream)
+            {
+                CV_Assert(mat.type() == CV_32F);
+                CV_Assert(mat.total() >= tensor.size());
+
+                Mat source = mat.isContinuous() ? mat : mat.clone();
+                CV_Assert(source.isContinuous());
+
+                memcpy<float>(reinterpret_cast<float*>(source.data), tensor.get(), tensor.size(), stream);
+
+                if (source.data != mat.data)
+                    source.copyTo(mat);
+            }
+        }} /* cuda4dnn::csl */
+
+        /* base class for all CUDA backend/target node */
+        class CUDABackendNode : public BackendNode {
+        public:
+            CUDABackendNode() : BackendNode(DNN_BACKEND_CUDA) { }
+            virtual ~CUDABackendNode() { }
+
+            virtual void forward(
+                const std::vector<cv::Ptr<BackendWrapper>>& inputs,
+                const std::vector<cv::Ptr<BackendWrapper>>& outputs,
+                cuda4dnn::csl::Workspace& workspace) = 0;
+
+            virtual std::size_t get_workspace_memory_in_bytes() const noexcept { return 0; }
+        };
+
+        /* utility function which creates a correct backend node based on `targetId` */
+        template <template <class> class NodeType, class ...Args>
+        cv::Ptr<BackendNode> make_cuda_node(int targetId, Args&& ...args)
+        {
+            switch (targetId)
+            {
+            case DNN_TARGET_CUDA_FP16:
+                return Ptr<BackendNode>(new NodeType<half>(std::forward<Args>(args)...));
```

Comment:
At first look, it feels like `makePtr` should be used. But `makePtr` doesn't work well with rvalues.

```
template<typename _Tp, typename ... A1> static inline
Ptr<_Tp> makePtr(const A1&... a1)
{
    static_assert( !has_custom_delete<_Tp>::value, "Can't use this makePtr with custom DefaultDeleter");
    return (Ptr<_Tp>)std::make_shared<_Tp>(a1...);
}
```

`makePtr` does not work well with temporaries or xvalues. This leads to unncessary copying.

Is there something like the following:

```
template<typename _Tp, typename ... A1> static inline
Ptr<_Tp> makePtr(A1&&.. a1)
{
    static_assert( !has_custom_delete<_Tp>::value, "Can't use this makePtr with custom DefaultDeleter");
    return (Ptr<_Tp>)std::make_shared<_Tp>(std::forward<A1>(a1)...);
}
```

References: https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared